### PR TITLE
Changes to F4 Serial RX code.

### PR DESCRIPTION
**Pull request by victorpv**
Changes to F4 serial code so incoming packets are not dropped when the buffer is full.
Now the cdc driver will not send acknowledge to the last packet until it can be written
to the buffer.
With this change the F4 works similar to the F1, when the buffer is full it sends NACKs rather than ACK.
Untli now the F4 would not notify the host the buffer is full and instead would just drop the packet, causing data loss.
Also includes some changes to CCM RAM to relocate the usb buffers, stack and heap there, similar to Steve's F4 branch.

### CODE REVIEW

**Comment by stevstrong** [comment]
Victor, can you tell me how to check the problem and the solution?

**Comment by victorpv** [comment]
The problem is that it drops bytes. You can send from the host to the board even if the sketch doesn't care to ever read from the buffer.
An empty sketch that opens the port with Serial.being, then sends something and blinks a led, should be enough to test. Then send data from a host terminal, and the host will never block waiting for the device to read. So if you have a sketch that needs to actually get data from the host, but is getting such data at a slower rate than the host can send it, will lose data.  For example if you are trying to send an image or something like that.
You can also test with some more complicated sketch. In my case I discovered it because I was trying to compare the speed with the F1, so I was using this test:
https://www.pjrc.com/teensy/benchmark_usb_serial_receive.html

During those tests I got incredible speeds, so I started to check whether the data was actually being received by the sketch that fast, and started by echoing back to the host, and found that I was losing bunchs of data, but the host would send as fast as it could nonetheless. Then I tested to just completely stop reading after the sketch had started, and found the host could still keep sending data at top speed, it would never stop or slow down, even though the sketch was not reading it at all.
Given the host will send packets to the USB serial port as fast as it can, even if you open the port at a certain speed (lets say 115200bps), with the current F4 code there is always the chance of losing characters if at any moment the buffer fills up. If you are sending commands, images or anything else, there is a potential for data loss.

In the F1, teensy, due, the USB driver in the board will stop sending acks if the RX buffer fills up, and when send acks again when there is space for at least one more packet (64bytes). That acts as a sort of flow control, so the host can't send faster than the sketch is reading the data.
In the arduino AVR ones, the usb to serial chip will process data at the rate of the usart connection to the AVR, so it will hold the host if the host is sending faster than whichever speed you open the port at in the Arduino. So it can drop data if the arduino doesn't read at least that fast, but doesn't run the risk of the host sending at the full USB port speed when the sketch can't read that fast.

**Comment by rogerclarkmelbourne** [comment]
Is this one ready to merged (Steve ??)

**Comment by stevstrong** [comment]
Sorry, I did not have the time yet to test, but I will test it next week.
This PR also adds the CCMRAM definitions to the core which is good. I have this already in my repo and works fine.

**Comment by rogerclarkmelbourne** [comment]
Steve

If you want, I'll pull your whole fork to my local machine into a different folder, and then copy your F4 section to replace my F4 folder.

Then at least the F4 will be up to date with your changes without any issues with the F1 code getting changed at the same time.

**Comment by stevstrong** [comment]
I think I have to first make a Processing sketch to send a lot of data over USB serial to F4 board in order to reproduce the problem, and then test the solution.

**Comment by stevstrong** [comment]
OK, now I followed the link give by Victor and downloaded the benchmark package and ran the receive_test.exe file.
I got "Average bytes per second = 624029", without any USB serial application running on the F4 board.
Which is indeed unrealistic.
So now I am able to test the solution. :o)
I will let you know about that.

**Comment by stevstrong** [comment]
Victor, I started a review for a possible issue, it would be nice if you could comment on that.

**Comment by victorpv** [comment]
Steve can you link to that review?

File: STM32F4/cores/maple/libmaple/usbF4/VCP/usbd_cdc_vcp.c [hunk 1]
```diff
@@ -51,31 +52,90 @@ extern volatile int APP_Rx_ptr_in;    /* Increment this pointer or roll it back
...
 
 uint8_t VCPGetByte(void) {
+    /*uint32 rx_unread = (UsbRecWrite - UsbRecRead) & UsbRecBufferSizeMask;
+    if ((rx_unread <= UsbRecBufferSize - CDC_DATA_MAX_PACKET_SIZE) && UsbHandle ){
+        usbd_cdc_PrepareRx(UsbHandle);
+    }
+    */
+
 	if(UsbRecWrite == UsbRecRead) {
 		return 0;
```
[the remark this answers is not in the document]
Comment (victorpv) [reply STM32F4/cores/maple/libmaple/usbF4/VCP/usbd_cdc_vcp.c]: I agree this should be addressed. I vaguely remember thinking about this one of some other function returning bytes, and there was no clear way to return when nothing is available. Then I looked at the Arduino code, and I can't remember what I found by now, but I think  Arduino either returned 0, or used an int (vs uint) and returned  -1 when no data was available.
I think using an int16_t as the return byte may be the best option, and return -1 when no byte is available, and 0 to 255 for actual data. I believe returned values are normally in register R0, and returning byte vs returning int16_t shouldn't make the code any slower as either one can fit in a single register.

[the remark this answers is not in the document]
Comment (stevstrong) [reply STM32F4/cores/maple/libmaple/usbF4/VCP/usbd_cdc_vcp.c]: I agree, returning int16 is an acceptable solution, just that the upper level functions should take care of that.

File: STM32F4/cores/maple/libmaple/usbF4/usb.c [hunk 2]
```diff
@@ -61,20 +65,24 @@ void usbDisableBlockingTx(void) {
...
+
+
 uint32_t usbReceiveBytes(uint8_t* recvBuf, uint32_t len) {
-	  int newBytes = usbBytesAvailable();
+	  int newBytes = VCPBytesAvailable();
 	  if (len > newBytes) {
 	      len = newBytes;
 	  }
+	  return VCPGetBytes(recvBuf, len);
+/*
```
[the remark this answers is not in the document]
Comment (stevstrong) [reply STM32F4/cores/maple/libmaple/usbF4/usb.c]: Yes, it seems reasonable.

**Comment by victorpv** [comment]
I only see 2 uses of VCPGetByte, and both would be more efficient using VCPGetBytes, and my PR changes at least one fo them to getBytes:
https://github.com/rogerclarkmelbourne/Arduino_STM32/search?utf8=%E2%9C%93&q=VCPGetByte&type=

So with that in mind, I think the best would be to remove VCPGetByte and leave only VCPGetBytes, and change the other call in the core to VCPGetBytes.

And then in VCPGetBytes, make the change I suggested and leave it only as:
uint32_t usbReceiveBytes(uint8_t* recvBuf, uint32_t len) {
return VCPGetBytes(recvBuf, len);
}

That should speed up so by removing redundant checks and eliminate the problem of getByte returning 0 for no data.

**Comment by victorpv** [comment]
@stevstrong any progress testing this?
If you want I can download your fork and send a PR for your fork.

**Comment by victorpv** [comment]
Sounds good. The problem happens because when the USB peripheral gets a packet it triggers an ISR That ISR checks if the is space in the buffer or not, if there is copy the packet there, if there isn't then doesn't copy the packet there. 
But along that process, it never cares to change the USB peripheral from acknowledge to not-acknowledge, so the USB peripheral acknowledges the packet to the host, and the host can submit a new one (even though the packet in the USB buffer may not have been copied out, so will be overwritten by the incoming packet).
In the libmaple code, the ISR will only let the peripheral acknowledge the packet if the packet was copied to out of the hardware buffer. Otherwise it will instead send NACKs to the host, and the host will keep repeating the same packet until eventually acknowledged. So the packet in the USB buffer will be overwritten too, but by the very same packet sent again by the host. That will cause  a new ISR, and the ISR will check again if there is space in the ram buffer to copy the packet. If this second time there is still no space, new NACK to the host, and host resends the packet again. If eventually the ISR finds space in the buffer, it will then send an ACK to the host, indicating the packet was received correctly, and the host will move on to the next packet in it's buffer.

Something like:
Host: send packet 1
F4: receive packet 1, can't buffer it, send NACK
Host: get NACK, resend packet 1
F4: receive packet 1, can't buffer it, send NACK
Host: get NACK, resend packet 1
F4: receive packet 1, this time can buffer it, send ACK
Host: get ACK, packet 1 was good, send packet 2.

In the original F4 code, the functions were set to always send ACK no matter what. The changes adds that handshaking process as per libmaple (and per the USB docs I could find)

**Comment by stevstrong** [comment]
I understand, just that I don't see yet which function sends ACK or NACK.

**Comment by victorpv** [comment]
This is the key part:
https://github.com/rogerclarkmelbourne/Arduino_STM32/pull/452/files#diff-96e3df7aff9d29c23d8445f6b2e2f26eR739

In the old code, it always prepares the endpoint to get the next packet, even before it tries to move data out of the hardware buffer.
In the new, it only prepares the endpoint to get the next packet if the packet was moved out of the HW buffer successfully.
Besides that, there is a new function that get's called if the sketch reads packet out of the buffer and makes enough space for a new packet. In that case the new function is called to set the enpoint to receive packets again,
The rest of changes are just to allow those 2 things to work.

**Comment by stevstrong** [comment]
Victor, I made some optimizations myself in the F4 USB part, and with these changes I cannot reproduce your problem.

Without using the key part you pointed out, I get 100% Rx data with 714kBps Rx throughput.
I am using the test files used that time for F1 (http://www.stm32duino.com/viewtopic.php?f=9&t=1451&hilit=image+loader&start=50#p19003) adapted for ILI9341, and I get the whole picture (76800 bytes) from the PC within 215 ms. No missed bytes.
I tested 10 consecutive times, still no missed bytes.

Do you have some test bench which I can use to reproduce the problem?
Or maybe I solved the issue through my changes? I only did code size & speed optimizations.

**Comment by victorpv** [comment]
The F4 is fast :)
From looking at the code a while back, there is a lot of optimizations we could do. I didn't want to make those changes at the same time as correcting the RX bug to keep a clear history of changes, in case we later had to revert something, but there is a lot of things that could be optimized to increase speed and reduce CPU usage.
Another trick the driver doesn't use is the USB doublebuffering. If you haven't read about it, in short since the CDC packets are only 64bytes,the hardware usb buffer could fit 2 buffers for RX and 2 for TX, alternating them when a packet comes in, so the end point goes straight to receiving another packet while the first packet is copied out of the hardware buffer. If I remember right the software has to flip a flag to switch between both buffers, but that would be fairly fast comparing having to wait until the single buffer data is moved out. The F1 can do the same thing. Although I am not sure how much faster that would make the port, since anyway every packet has to be copied, so once the second packet arrives if the first is still being copied, it will have to wait. Anyway, there are other possible optimizations, and I'm sure you will see many more than I could see, thighter loops, less jumping around, faster ways to copy the data...

**Comment by stevstrong** [comment]
Yea, the bottleneck is the data copying, and if it can be done within two consecutive Rx frames then double buffering does not bring any advantage.
Would be happy to check any further optimizations you suggest.
But now I go back and test your fix  :)

**Comment by victorpv** [comment]
Yeah let's complete the fix and we can try optimizations later.
If I remember right from the time I tested, with the fix I was still getting somewhere around 700KB, which made me think the difference with the F1 was due to the cpu speed since it was about double. But this is all from memory.
I have noticed the way you copy multiple bytes within a single loop run in the SDIO code, that would probably help here too.

**Comment by victorpv** [comment]
Great!
The other day when going thru the code to point to your the important parts, I noticed I was not checking whether the endpoint was enabled before re-enabling, but I wrote this several months ago and worked so I thought it probably wasn't necessary. I glad you for corrected it.
Did you merge it to your repo? If so I'll go have a look to see what else you changed.

**Comment by victorpv** [comment]
Did you get rid of vcpGetByte?
If so, this line can be removed:
https://github.com/stevstrong/Arduino_STM32/commit/2f65f256c36e5300bef232299b439706c3ee063b#diff-11e49d9ed80570e833a2d2dd4cce576dR38

**Comment by stevstrong** [comment]
That's right, that chunk of `VCPGetByte` can be removed. I realized that is enough to have the `VCPGetBytes()`, it handles also the single byte read.

The declaration of `usbd_cdc_PrepareRx` can be of course placed anywhere else, but as long as it is not used elsewhere, I though that place is also fine, it does not confuse reading the rest of the code, and most important it compiles and works correctly. Should there be a need to use it on other places, too, it can be moved as necessary.

**Comment by victorpv** [comment]
Can't object to that. I'll try to test your version as soon as I get a chance and let you know if I have any problem.
If I understand right, during your test you are getting over 800KBps, which is even higher than what you got before the fixes with the code that drop data right?
I think that's impresive :)

**Comment by stevstrong** [comment]
In my test I get an average of ~800kBps, 825 was the max I achieved. But it strongly depends on the host (PC). If the defender runs in background on my Win10 machine then it slows down, otherwise I get  higher speeds.

This is indeed higher than before due to some optimizations I made, for instance load/store the volatile variables so that the buffer index (which gets incremented and ANDed by the buffer size mask) is a register variable instead of a volatile.
The other improvement was the replacement of single byte read with buffered read, as you have it in this PR.

**Comment by victorpv** [comment]
> for instance load/store the volatile variables so that the buffer index (which gets incremented and ANDed by the buffer size mask) is a register variable instead of a volatile.

If I understand right, how that works if that if you updating the volatile in each loop run, it needs to write to the memory because is volatile, but if you load the volatile to a temp variable, then at the end of the loop you load that back to the volatile, the RAM is only accessed twice (read at start and write at finish) and the loop instead just use a CPU register. Is that right?

**Comment by stevstrong** [comment]
Exactly. The compiler may not optimize the volatile accesses, but can (and hopefully will) optimize the register accesses.

**Comment by victorpv** [comment]
Thanks for explaining, didn't know that trick.

**Comment by stevstrong** [comment]
I have to test this again using your repo, me I have lately massively reworked the CDC USB part of my F4 repo, including a fix for Tx.
So i eventually will send you the involved complete files, as you know my PR preparing knowledge is very limited.

**Comment by rogerclarkmelbourne** [comment]
OK.

Thanks.

**Comment by stevstrong** [comment]
It does not make sense to merge this since:
- it includes a lot of changed files which could eventually have negative effect on the whole F4 branch
- [my repo](https://github.com/stevstrong/Arduino_STM32) offers a better supported framework to work with F4.
Should rise the need to have this repo for F4 up-to-date, then the actual version of my F4 repo could be eventually merged here.

**Comment by fpistm** [comment]
> * Should rise the need to have this repo for F4 up-to-date, then the actual version of my F4 repo could be eventually merged here.

So, you plan to mege it ?🤗
Seems @rogerclarkmelbourne promoted you as a maintainer 😉

**Comment by stevstrong** [comment]
Not yet. I think your (STM) repo also offers a good framework for F4, so I don't really see currently a need to do that. But who knows about the future? :)

**Comment by fpistm** [comment]
Right 😁
Future is unpredictable 🙄
Anyway, happy to see all this activities on this great core. And congrats for the promotion and the great work/advice you do 👏

**Pull request closed without merging**